File: Year 2/Algorithms/Recurrence relations/0.6/main.py
with open("input.txt", "r") as file:
    s = file.readline()
f = [[0] * (len(s)) for _ in range(len(s))]
for i in range(len(s)):
    f[i][i] = 1

for d in range(1, len(s)):
    for i in range(len(s) - d):
        j = i + d;
        if s[i] == s[j]:
            f[i][j] = f[i + 1][j - 1] + 2
        else:
            f[i][j] = max(f[i + 1][j], f[i][j - 1])
       
left = []
right = []
i = 0
j = len(s) - 1
while f[i][j] != 0:
    if  s[i] == s[j]:
        left.append(s[i])
        i += 1
        j -= 1
    elif f[i][j] == f[i + 1][j]:
        i += 1
    elif f[i][j] == f[i][j - 1]:
        j -= 1

for i in range(f[0][len(s) - 1] // 2 - 1, -1, -1):
    right.append(left[i])
pal = ''.join(left + right)

with open("output.txt", "w") as file:
    file.write(str(f[0][len(s) - 1]) + "\n")
    file.write(str(pal))

File: Year 2/Algorithms/Recurrence Relations/0.6/main.cpp
#include <fstream>
#include <string>
#include <vector>

long long myMax(long long a, long long b)
{
    return a > b ? a : b;
}

int main() {
    std::ifstream fin("input.txt");
    std::ofstream fout("output.txt");

    std::string s;
    std::getline(fin, s);
    int n = s.size();

    std::vector<std::vector<int>> f(n, std::vector<int>(n, 0));

    for (int i = 0; i < n; ++i) f[i][i] = 1;

    for (int d = 1; d < n; ++d)
    {
        for (int i = 0; i < n - d; ++i)
        {
            int j = i + d;
            if (s[i] == s[j])
            {
                f[i][j] = f[i + 1][j - 1] + 2;
            }
            else
            {
                f[i][j] = myMax(f[i + 1][j], f[i][j - 1]);
            }
        }
    }

    std::string left = "", right = "";
    int i = 0, j = n - 1;
    while (i <= j)
    {
        if (s[i] == s[j])
        {
            left += s[i];
            if (i != j) right = s[j] + right;
            ++i;
            --j;
        }
        else if (i + 1 < n && f[i + 1][j] >= f[i][j - 1])
        {
            ++i;
        }
        else
        {
            --j;
        }
    }
    std::string pal = left + right;

    fout << f[0][n - 1] << std::endl;
    fout << pal << std::endl;

    return 0;
}